tf2_ros/include/tf2_ros/buffer.h

```diff
@@ -262,6 +262,7 @@ namespace tf2_ros
 
     /// \brief A map from active timers to BufferCore request handles
     std::unordered_map<TimerHandle, tf2::TransformableRequestHandle> timer_to_request_map_;
+    std::unordered_map<TimerHandle, tf2::TransformableCallbackHandle> timer_to_callback_map_;
```

Comment:
Unfortunately, this still breaks ABI, by adding a new member to a public class.

One ugly way to fix this that I've used in the past is to put a global variable into the `buffer.cpp` file which is a map of objects to maps.  Then when you need to look this up, you first look up the correct map by the object pointer address, and then you can look up the `tf2::TransformableCallbackHandle` from that map.  This maintains ABI, while still preserving the information you want.  Note that since this is a global map, it will also need locking since multiple threads would need to be in here.

So to summarize, the variables you would need to add at the top of `buffer.cpp` would be something like:

```
std::mutex g_object_map_to_cb_handle_mutex;
std::map<void*, std::unordered_map<TimerHandle, tf2::TransformableCallbackHandle>> g_object_map_to_cb_handle;
```

And then when you go to use this, you take the mutex, use "this" to look up the correct map, then use the TimerHandle in `this` to lookup the `TransformableCallbackHandle`.

You can see an example of this pattern in https://github.com/ros2/rclcpp/pull/1056

Reply:
Does adding a buffer destructor maintain ABI? If not, how do you propose I remove objects `g_object_map_to_cb_handle`?

Reply:
Apparently adding one can change the ABI: https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html .  So that doesn't seem to be an option.

How about when we remove items from the map, we check to see if the map at that particular address is empty?  If it is, then we remove the map.  This will also require other parts of the code to check to see if the map exists (and create it if it doesn't), but should maintain ABI.

Reply:
That is not what I mean. I think having the map all the time is fine but not removing mappings to destroyed Buffer objects introduces a new memory leak. In ros2/rclcpp#1056 the mapping from the object to the list is removed from `g_object_map_to_cb_handle` in the destructor but we can't do that here. How do you propose we remove mappings in this MR?

Reply:
I don't have a good idea off-hand.  The tools we have to work with are adding new (non-virtual) class methods, new free functions, and new global variables.  We'd have to do something with that combination of tools to make this work.